### PR TITLE
feat: add OpenSearchProvider and OpenSearchSessionMapper

### LINKED ISSUE #191: [FEATURE] Add OpenSearchProvider for evaluating traces stored in OpenSearch
## Problem Statement

There is currently no way to run strands-evals against agent traces stored in OpenSearch. This affects users of [observability-stack](https://github.com/opensearch-project/observability-stack), [agent-health](https://github.com/opensearch-project/agent-health), Amazon OpenSearch Service, or anyone else using OpenSearch as their storage backend for OTel GenAI agent traces.

## Proposed Solution

Add `OpenSearchProvider(TraceProvider)` and `OpenSearchSessionMapper(SessionMapper)`, same pattern as [CloudWatchProvider](https://github.com/strands-agents/evals/blob/main/src/strands_evals/providers/cloudwatch_provider.py) and [LangfuseProvider](https://github.com/strands-agents/evals/blob/main/src/strands_evals/providers/langfuse_provider.py).

**OpenSearchProvider**
- Wraps [`OpenSearchTraceRetriever`](https://github.com/opensearch-project/genai-observability-sdk-py) from `opensearch-genai-observability-sdk-py` for querying and auth (basic, SigV4, none)
- Implements `get_evaluation_data(session_id)` returning `TaskOutput`
- Queries by conversation ID or trace ID (retriever handles both)

**OpenSearchSessionMapper**
- Converts genai-sdk `SpanRecord` objects to strands-evals `Session`/`Trace`/`Span` types
- Maps `invoke_agent` to `AgentInvocationSpan`, `execute_tool` to `ToolExecutionSpan`, `chat` to `InferenceSpan`
- Scopes tool attribution to parent agent spans for correct multi-agent behavior

**Dependency**: `opensearch-genai-observability-sdk-py[opensearch]>=0.2.7` as optional extra (`pip install strands-agents-evals[opensearch]`)

### Local / development (basic auth)

```python
from strands_evals.providers import OpenSearchProvider
from strands_evals.evaluators import HelpfulnessEvaluator

provider = OpenSearchProvider(
    host="https://localhost:9200",
    auth=("admin", "password"),
    verify_certs=False,
)

task_output = provider.get_evaluation_data(session_id="my-session")
evaluator = HelpfulnessEvaluator(model="us.anthropic.claude-sonnet-4-20250514-v1:0")
results = evaluator.evaluate(task_output)
```

### Amazon OpenSearch Service (SigV4)

```python
from opensearchpy import RequestsAWSV4SignerAuth
import boto3

credentials = boto3.Session().get_credentials()
auth = RequestsAWSV4SignerAuth(credentials, "us-east-1", "es")
provider = OpenSearchProvider(
    host="https://my-domain.us-east-1.es.amazonaws.com",
    auth=auth,
)
```

## Alternatives Considered

Could use `opensearch-py` directly instead of wrapping the genai-sdk (similar to how CloudWatchProvider uses boto3). The genai-sdk was chosen because it already handles Data Prepper's de-dotted field mappings and OTel message parsing, which is non-trivial to reimplement, plus auth handling (basic, SigV4 signing for AWS managed OpenSearch). Open to the direct approach if maintainers prefer fewer transitive dependencies.

## Prior Art

- [CloudWatchProvider](https://github.com/strands-agents/evals/pull/147)
- [LangfuseProvider](https://github.com/strands-agents/evals/pull/144)
- [TraceProvider interface](https://github.com/strands-agents/evals/pull/140) ([#97](https://github.com/strands-agents/evals/issues/97))

## Implementation

PR: #192

### PR DESCRIPTION
**Pull request by kylehounslow**
## Description

Add OpenSearch as a trace provider for evaluations. This enables users storing agent traces in OpenSearch (local or Amazon OpenSearch Service) to retrieve them and run evaluators against them.

- `OpenSearchProvider(TraceProvider)` queries OpenSearch for spans by session/trace ID, returns `TaskOutput`
- `OpenSearchSessionMapper(SessionMapper)` converts span documents to strands-evals `Session`/`Trace`/`Span` types
- `opensearch` optional dependency group (`pip install strands-agents-evals[opensearch]`)
- Lazy-loaded provider (same pattern as CloudWatch/Langfuse)

Uses [opensearch-genai-observability-sdk-py](https://github.com/opensearch-project/genai-observability-sdk-py) for OpenSearch querying, auth (basic, SigV4), and span field mapping.

## Related Issues

Closes #191

## Type of Change

New feature

## Testing

- 32 unit tests covering mapper, provider, edge cases, multi-agent tool scoping, and large traces
- Passes `hatch run test-lint` (ruff + mypy)

<details>
<summary>E2E verification with Strands agent traces</summary>

Sent a Strands agent request (Bedrock Claude) with traces exported to OpenSearch via OTel Collector + Data Prepper. Verified `OpenSearchProvider.get_evaluation_data()` correctly returns typed spans:

| Backend | Auth | Result |
|---------|------|--------|
| OpenSearch (local) | Basic auth | ✅ AgentInvocationSpan + InferenceSpan with correct prompt/response |
| Amazon OpenSearch Service | SigV4 | ✅ Multi-agent trace: 4 AgentInvocationSpans + 7 ToolExecutionSpans |

Full eval loop — Strands agent traces retrieved from OpenSearch and scored:

```
>>> provider = OpenSearchProvider(host="https://localhost:9200", auth=("admin", "pw"), verify_certs=False)
>>> experiment = Experiment(cases=cases, evaluators=[HelpfulnessEvaluator(...)])
>>> reports = experiment.run_evaluations(task)

evaluator_name='HelpfulnessEvaluator' overall_score=0.833 test_passes=[True]
```

</details>

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [x] I have updated the documentation accordingly
- [x] My changes generate no new warnings

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -45,6 +45,7 @@ dev = [
 ]
 
 langfuse = ["langfuse>=2.0.0,<4"]
+opensearch = ["opensearch-genai-observability-sdk-py[opensearch]>=0.2.7"]
```
Comment (afarntrog) [inline pyproject.toml]: let's put an upper bound on this dep

File: src/strands_evals/providers/opensearch_provider.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        return TaskOutput(output=output, trajectory=session)
+
+    @staticmethod
+    def _extract_output(session):
+        """Extract final agent response from the last AgentInvocationSpan."""
+        for trace in reversed(session.traces):
+            for span in reversed(trace.spans):
+                if isinstance(span, AgentInvocationSpan) and span.agent_response:
+                    return span.agent_response
+        return ""
```
Comment (afarntrog) [inline src/strands_evals/providers/opensearch_provider.py:104]: nit: out of scope for this pr but all three providers (`CloudWatch`, `Langfuse`, `OpenSearch`) now have an identical `_extract_output` method. We should consider extracting it to the `TraceProvider` base class or a helper function in that module.

**Review by afarntrog — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (kylehounslow) [reply pyproject.toml]: Good call. Added `<1.0.0` as upper bound since this SDK is pre-1.0

File: src/strands_evals/providers/opensearch_provider.py (hunk 2 above)
Comment (kylehounslow) [reply src/strands_evals/providers/opensearch_provider.py:104]: Good catch, makes sense as a future cleanup.

**Review by afarntrog — APPROVED** [review]
